### PR TITLE
Fixes 40012 "Malformed message; invalid clientId" when message has no clientId and credentials can assume any clientId

### PR DESCRIPTION
**Pull request by ricardopereira**
#1074

### CODE REVIEW

**Comment by ricardopereira** [comment]
Screenshot 2020-10-23 at 09 26 56

File: Source/ARTHttp.m [hunk 1]
```diff
@@ -47,14 +47,22 @@ - (void)dealloc {
 }
 
 - (NSObject<ARTCancellable> *)executeRequest:(NSMutableURLRequest *)request completion:(void (^)(NSHTTPURLResponse *_Nullable, NSData *_Nullable, NSError *_Nullable))callback {
-    [self.logger debug:@"--> %@ %@\n  Body: %@\n  Headers: %@", request.HTTPMethod, request.URL.absoluteString, [[NSString alloc] initWithData:request.HTTPBody encoding:NSUTF8StringEncoding], request.allHTTPHeaderFields];
+    NSString *requestBodyStr = [[NSString alloc] initWithData:request.HTTPBody encoding:NSUTF8StringEncoding];
+    if (!requestBodyStr) {
+        requestBodyStr = [request.HTTPBody base64EncodedStringWithOptions:NSDataBase64EncodingEndLineWithCarriageReturn];
+    }
+    [self.logger debug:@"--> %@ %@\n  Body: %@\n  Headers: %@", request.HTTPMethod, request.URL.absoluteString, requestBodyStr, request.allHTTPHeaderFields];
```
Comment (QuintinWillison) [inline Source/ARTHttp.m:54]: This new code adds runtime overhead that's required for normal function. Do we have the capability to detect at runtime if debug logs are enabled and, therefore, only construct this string in that case?

Also, why can the call to `initWithData:encoding:` fail? And if it can fail, would it not be better to encapsulate this logic (used twice here) in to a function to perhaps document why it can fail in the one place?

File: Source/ARTRestChannel.m [hunk 2]
```diff
@@ -247,9 +247,6 @@ - (void)internalPostMessages:(id)data callback:(void (^)(ARTErrorInfo *__art_nul
             callback([ARTErrorInfo createWithCode:ARTStateMismatchedClientId message:@"attempted to publish message with an invalid clientId"]);
             return;
         }
-        else {
```
Comment (QuintinWillison) [inline Source/ARTRestChannel.m:250]: Could you explain somewhere in this PR, or against the original issue even, why this change was necessary and why it fixes the issue?

**Review by QuintinWillison — CHANGES_REQUESTED** [review]
I've not looked in detail at the tests as I'm short for time, but I've made some comment on the implementation.

File: Source/ARTHttp.m (hunk 1 above)
Comment (ricardopereira) [reply Source/ARTHttp.m:54]: @QuintinWillison I agree, it's necessary overhead for the normal use case. I'll find out a solution. `initWithData:encoding:` can fail when the data is encoded with MsgPack and that's why it should at least show a B64 string instead of null. When I was debugging I thought: `Oh, here's the problem, the body is empty` and that was not the case 😄

File: Source/ARTRestChannel.m (hunk 2 above)
Comment (ricardopereira) [reply Source/ARTRestChannel.m:250]: I think the removed code existed because of a misinterpretation of the spec. The library should not change the clientId. The clientId is explicitly configured with the `ClientOptions` or implicitly through Token Auth.

**Review by tiholic — APPROVED** [review]
`Source/ARTRestChannel.m` looks good to me.
Haven't looked at other changes.

**Review by QuintinWillison — APPROVED** [review]
